tox.ini

```
@@ -1,20 +1,100 @@
 [tox]
-envlist = py38, py39, py3.10, py311, py312, pypy3
+envlist = pypy3,py3{9,10,11,12,13}-{linux,macos,windows}
 skip_missing_interpreters = true
+isolated_build = true
+skipsdist = true
 
 [gh-actions]
 python =
-    3.8: py38
     3.9: py39
     3.10: py310
     3.11: py311
     3.12: py312
-    pypy3.10: pypy3
+    3.13: py313
+
+[gh-actions:env]
+PLATFORM =
+    ubuntu-latest: linux
+    ubuntu-22.04: linux
+    ubuntu-24.04-arm: linux
+    macos-latest: macos
+    macos-15: macos
+    windows-latest: windows
+    windows-2025: windows
 
 [testenv]
-extras = dev
+skip_install = true
+
+passenv =
+    DISPLAY
+    XAUTHORITY
+    HOME
+    USERNAME
+    USER
+    XDG_*
+    CI
+    OS
+    PYTHONIOENCODING
+    PIP_DOWNLOAD_CACHE
+
+setenv =
+    PYTHONPATH = {toxinidir}
+
+allowlist_externals =
+    bash
+
+deps =
+    -e .[test,dev]
+
+commands =
+    python -m pytest {posargs:}
+    python -m pylint cpplint.py
+    python -m flake8 cpplint.py
+
+[testenv:build]
+skip_install = true
+
+passenv =
+    pythonLocation
+    CI
+    PYTHONIOENCODING
+    PIP_DOWNLOAD_CACHE
+
+deps =
+    pip>=21.1
+    build
+    twine
+
+commands =
+    python -m build .
+    twine check dist/*
+
+[testenv:check]
+skip_install = true
+allowlist_externals =
+    bash
+
+passenv =
+    CI
+    GITHUB*
+    PIP_DOWNLOAD_CACHE
+
+deps =
+    pip>=21.1
+
+changedir =
+    dist/
+
+commands =
+    python -m pip install cpplint  --force --pre --prefer-binary -f ./
+
+[testenv:clean]
+skip_install = true
+allowlist_externals =
+    bash
+
+deps =
+    pip>=21.1
 
 commands =
-  {envpython} -m pytest {posargs:}
-  {envpython} -m pylint cpplint.py
-  {envpython} -m flake8 cpplint.py
+    bash -c 'rm -rf build/ dist/ __pycache__/ *.egg-info/ .coverage'
```

Is there a reason to specifically use bash here? Why not
```suggestion
    rm -rf build/ dist/ __pycache__/ *.egg-info/ .coverage
```

---

In this case either bash or rm needs to be in the allow list. Bash is usually my default "allow" and that gets used to run any other shell commands. In this it's just "rm".

---

GitHub workflows have an allowlist for commands?